frontend/src/components/course-search/CourseSearch.tsx

```diff
@@ -87,25 +92,74 @@ export default memo(function CourseSearch() {
         conflictingSectionsOptions,
     ]);
 
+    const { enableHistoricalSearch, historicalSearchRange } = useStore(
+        (store) => store.experimentalFeaturesOptions,
+    );
+    const allTerms = useAllTerms() ?? [];
+    const range = Math.min(allTerms.length, historicalSearchRange);
+    const historicalSections = useSectionsForTermsQuery(
+        enableHistoricalSearch && sectionsToShow?.length === 0,
+        allTerms.slice(0, range),
+    ).data;
+
+    const matchingHistoricalSections: APIv4.Section[] | undefined =
+        React.useMemo(() => {
+            if (historicalSections === undefined) return undefined;
+
+            const filteredHistoricalSections =
+                searchFilters.length === 0
+                    ? historicalSections
+                    : historicalSections.filter((s) =>
+                          Search.filterSection(
+                              s,
+                              searchFilters.map((s) => s.filter),
+                          ),
+                      );
+
+            if (searchText === "") return filteredHistoricalSections;
+
+            let res: [number, APIv4.Section][] = [];
+            for (const s of filteredHistoricalSections) {
+                const score = Search.matchesText(searchText, s, areas);
+                if (score !== null) {
+                    res.push([score, s]);
+                }
+            }
+            const sorted = res.sort((a, b) => b[0] - a[0]);
+            return sorted.map((a) => a[1]);
+        }, [
+            historicalSections,
+            searchText,
+            searchFilters,
+            enableHistoricalSearch,
+            historicalSearchRange,
+        ]);
+
     return (
         <div className={Css.container}>
             <SearchControls />
-            {sectionsToShow !== undefined ? (
-                <CourseSearchResults
-                    sections={sectionsToShow}
-                    searchKey={btoa(searchText)}
-                />
-            ) : (
-                <CourseSearchEnd text="loading courses..." />
-            )}
+            <div className={Css.resultsContainer}>
+                {sectionsToShow !== undefined ? (
+                    <CourseSearchResults
+                        sections={sectionsToShow}
+                        searchKey={btoa(searchText)}
+                    />
+                ) : (
+                    <CourseSearchEnd text="loading courses..." />
+                )}
+
+                {enableHistoricalSearch && sectionsToShow?.length === 0 ? (
+                    <HistoricalSearchResults
+                        sections={matchingHistoricalSections}
+                    />
+                ) : (
+                    <></>
+                )}
```

Comment:
This should be modified to only perform the historical search if the matched section count is 0 (and it's probably better to make it less than 5 or something like that). Currently your code only renders the result if there is no matched section, but performs the search unconditionally.

Reply:
I agree to have the historicalSearching done if the matching sections from current terms are less than some threshold (5 sounds good), and the performance issue will be handled like comment above.